### PR TITLE
add validation for MultiClusterService

### PR DESCRIPTION
**Pull request by yike21**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
add validation for MultiClusterService

**Which issue(s) this PR fixes**:
Part of #3749

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
NONE

```release-note
`karmada-webhook`: Add validation for `MultiClusterService` objects.
```

### CODE REVIEW

File: pkg/util/lifted/validatingmcs.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import (
+	"fmt"
+	"regexp"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	netutils "k8s.io/utils/net"
+)
+
+// +lifted:source=https://github.com/kubernetes/apimachinery/blob/v0.26.2/pkg/util/validation/validation.go#L202-L203
```
Comment (RainbowMango) [inline pkg/util/lifted/validatingmcs.go:32]: Can we just import this from `kubernetes/apimachinery`? 

The reason why we introduce and manage `lifted` code is we don't want to import `kubernetes/kubernetes` repo.

Comment (yike21) [reply pkg/util/lifted/validatingmcs.go:32]: I understand it and I will fix it soon :)

File: pkg/webhook/multiclusterservice/validating.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	klog.Infof("Validating MultiClusterService(%s/%s) for request: %s", mcs.Namespace, mcs.Name, req.Operation)
+
+	if req.Operation == admissionv1.Update {
+		oldMcs := &networkingv1alpha1.MultiClusterService{}
+		err = v.decoder.DecodeRaw(req.OldObject, oldMcs)
+		if err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		if errs := v.validateMCSUpdate(oldMcs, mcs); len(errs) != 0 {
+			klog.Errorf("%v", errs)
```
Comment (whitewindmills) [inline pkg/webhook/multiclusterservice/validating.go:44]: could you add some effictive logs? or just remove this.

File: pkg/webhook/multiclusterservice/validating.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+		apimachineryvalidation.NameIsDNS1035Label, field.NewPath("metadata"))
+	allErrs = append(allErrs, v.validateMultiClusterServiceSpec(mcs)...)
+	return allErrs
+}
+
+// validateMultiClusterServiceSpec validates MultiClusterService spec.
+func (v *ValidatingAdmission) validateMultiClusterServiceSpec(mcs *networkingv1alpha1.MultiClusterService) field.ErrorList {
+	allErrs := field.ErrorList{}
+	specPath := field.NewPath("spec")
+	var allPortNames map[string]bool
```
Comment (whitewindmills) [inline pkg/webhook/multiclusterservice/validating.go:81]: seems it's better to use `sets.Set`

Comment (yike21) [reply pkg/webhook/multiclusterservice/validating.go]: Thanks, it is more reasonable to use `sets.Set`.
Done.

File: charts/karmada/templates/_karmada_webhook_configuration.tpl [hunk 4]
```diff
@@ -199,4 +199,18 @@ webhooks:
...
+        apiVersions: ["*"]
+        resources: ["multiclusterservices"]
+        scope: "Namespaced"
+    clientConfig:
+      url: https://{{ $name }}-webhook.{{ $namespace }}.svc:443/validate-multiclusterservice
+      {{- include "karmada.webhook.caBundle" . | nindent 6 }}
+    failurePolicy: Fail
+    sideEffects: None
+    admissionReviewVersions: [ "v1" ]
+    timeoutSeconds: 3
```
Comment (jwcesign) [inline charts/karmada/templates/_karmada_webhook_configuration.tpl:215]: here is 3s, `artifacts/deploy/webhook-configuration.yaml` is 10s, can we keep them same?

Comment (yike21) [reply charts/karmada/templates/_karmada_webhook_configuration.tpl:215]: I'm also a little confused here, for the `timeoutSeconds` of webhook configuration in [charts/karmada/templates/_karmada_webhook_configuration.tpl](https://github.com/karmada-io/karmada/pull/3779/files/b5280dcf71280b1ad5a90cb9dec0c31d55cdf9c7#diff-18e1d51f488df0df7ce80b3131d3445be73a31bde39a7c36ce42c71e940a8a47) is 3s, so I make it 3s here.
And we can see that `timeoutSeconds` is different among configuration files, such as `artifacts/deploy/webhook-configuration.yaml`, `charts/karmada/templates/_karmada_webhook_configuration.tpl`.

Comment (jwcesign) [reply charts/karmada/templates/_karmada_webhook_configuration.tpl:215]: Let's set all of them to 3s. Is that ok? @chaunceyjiang @XiShanYongYe-Chang

Comment (yike21) [reply charts/karmada/templates/_karmada_webhook_configuration.tpl:215]: In #2237 , the timeout duration is extended to 10s. It's been a while, maybe we can first see what happens when the timeout duration is changed to 3s. If the problem with the CI environment has been fixed, we keep the timeout duration 3s. And if the CI has frequent errors later on, we can then set the timeout duration more leniently. Is this ok? @XiShanYongYe-Chang @jwcesign @chaunceyjiang

Comment (XiShanYongYe-Chang) [reply charts/karmada/templates/_karmada_webhook_configuration.tpl:215]: ok. I agree with it, thx.

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks~
/lgtm

**Review by RainbowMango — APPROVED** [review]
/approve

**Comment by XiShanYongYe-Chang** [comment]
Hi @yike21, a webhook call timeout occurs in someone's pr: https://github.com/karmada-io/karmada/actions/runs/5549370069/jobs/10133491595?pr=3788, can you help analyze it?

We can get the log here:

image

If necessary, we can create a new issue to track it.

**Comment by yike21** [comment]
> Hi @yike21, a webhook call timeout occurs in someone's pr: https://github.com/karmada-io/karmada/actions/runs/5549370069/jobs/10133491595?pr=3788, can you help analyze it?
> 
> We can get the log here:
> 
> image
> 
> If necessary, we can create a new issue to track it.

Yes, I will try to analyze it.

**Comment by chaunceyjiang** [comment]
I also try to analyze it.

**Pull request merged**